lib/bap_future/bap_future.ml

```diff
@@ -0,0 +1,516 @@
+open Core_kernel.Std
+
+module Std = struct
+
+  module Promise = struct
+    type 'a t = {
+      made: 'a -> unit;
+      filled: bool ref;
+    }
+
+    let create made = {made; filled = ref false;}
+
+    let fulfill p x = 
+      p.made x;
+      p.filled := true
```

Comment:
can you guarantee, that `fulfill` is the only method to fulfill the future? The `bind` method will fulfill it directly, and your mechanism will break.